api: Introduce & expose endpoint controller statuses

tgraf opened the pull request:
```
$ cilium status --all-controllers
[...]
Controller Status (0/2 failing)
  Name                               Last success   Last error   Count   Message
  sync-identity-to-k8s-pod (56326)   36s ago        never        0       no error
  sync-identity-to-k8s-pod (29898)   32s ago        never        0       no error
```

```
    $ cilium status -o json
      "controllers": [
        {
          "configuration": {
            "error-retry": true,
            "interval": "1m0s"
          },
          "name": "sync-identity-to-k8s-pod (56326)",
          "status": {
            "last-failure-timestamp": "0001-01-01T00:00:00.000Z",
            "last-success-timestamp": "2018-02-05T17:43:34.099-08:00",
            "success-count": 1
          },
          "uuid": "247d5d37-0adf-11e8-aa78-080027559fd9"
        },
        {
          "configuration": {
            "error-retry": true,
            "interval": "1m0s"
          },
          "name": "sync-identity-to-k8s-pod (29898)",
          "status": {
            "last-failure-timestamp": "0001-01-01T00:00:00.000Z",
            "last-success-timestamp": "2018-02-05T17:43:37.997-08:00",
            "success-count": 1
          },
          "uuid": "26d05710-0adf-11e8-aa78-080027559fd9"
        }
```

tgraf committed:
api: Introduce & expose endpoint controller statuses

Signed-off-by: Thomas Graf <thomas@cilium.io>

ianvernon commented on api/v1/openapi.yaml:
```diff
@@ -1407,6 +1410,60 @@ definitions:
+            description: Total number of successful runs
+            type: integer
+          last-success-timestamp:
+            description: Timestsamp of last success
```
Typo: timestamp

tgraf replied:
fixed

ianvernon commented on pkg/controller/controller.go:
```diff
@@ -84,15 +86,17 @@ type ControllerParams struct {
+	lastErrorStamp    time.Time
+	uuid              string
+	stop              chan struct{}
+	mutex             lock.RWMutex
```
In other structures we have put the mutex above all of the resources which should be accessed under that lock - can you do that here?

tgraf replied:
fixed

ianvernon reviewed (approved):
LGTM apart from some minor nitpicks.

tgraf committed:
controller: Split code into manager.go

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
controller: Provie global controller status in agent health

The global controller health will become the standard health overview of the
agent as it will show the status of all async operations:

$ cilium status --all-controllers
[...]
Controller Status (0/2 failing)
  Name                               Last success   Last error   Count   Message
  sync-identity-to-k8s-pod (56326)   36s ago        never        0       no error
  sync-identity-to-k8s-pod (29898)   32s ago        never        0       no error

$ cilium status -o json
  "controllers": [
    {
      "configuration": {
        "error-retry": true,
        "interval": "1m0s"
      },
      "name": "sync-identity-to-k8s-pod (56326)",
      "status": {
        "last-failure-timestamp": "0001-01-01T00:00:00.000Z",
        "last-success-timestamp": "2018-02-05T17:43:34.099-08:00",
        "success-count": 1
      },
      "uuid": "247d5d37-0adf-11e8-aa78-080027559fd9"
    },
    {
      "configuration": {
        "error-retry": true,
        "interval": "1m0s"
      },
      "name": "sync-identity-to-k8s-pod (29898)",
      "status": {
        "last-failure-timestamp": "0001-01-01T00:00:00.000Z",
        "last-success-timestamp": "2018-02-05T17:43:37.997-08:00",
        "success-count": 1
      },
      "uuid": "26d05710-0adf-11e8-aa78-080027559fd9"
    }

Signed-off-by: Thomas Graf <thomas@cilium.io>

raybejjani commented on pkg/client/client.go:
```diff
@@ -125,8 +127,21 @@ func formatNodeAddress(w io.Writer, elem *models.NodeAddressingElement, title, p
+
 // FormatStatusResponse writes a StatusResponse as a string to the writer
-func FormatStatusResponse(w io.Writer, sr *models.StatusResponse) {
+func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, allControllers bool) {
```
Could you docuement what `allControllers` does?

tgraf replied:
fixed

raybejjani commented on pkg/client/client.go:
```diff
@@ -199,4 +214,38 @@ func FormatStatusResponse(w io.Writer, sr *models.StatusResponse) {
+	if sr.Controllers != nil {
+		nFailing, out := 0, []string{"  Name\tLast success\tLast error\tCount\tMessage\n"}
+		for _, ctrl := range sr.Controllers {
+			if status := ctrl.Status; status != nil {
```
Since we only do something when `status != nil`, we can continue when `status == nil` and avoid the indent.

tgraf replied:
fixed

raybejjani commented on pkg/client/client.go:
```diff
@@ -199,4 +214,38 @@ func FormatStatusResponse(w io.Writer, sr *models.StatusResponse) {
+					}
+				} else {
+					nFailing++
+				}
```
I found this logic a bit difficult. Why not:
```golang
if status.ConsecutiveFailureCount > 0 {
	nFailing++
} else if !allControllers {
	continue
}
```

or a switch on the same clauses.

tgraf replied:
fixed

raybejjani reviewed (approved):
LGTM
